File: core/src/main/java/javabot/operations/time/Tri.java
package javabot.operations.time;

public class Tri<T> {
    private static int OFFSET = 'a';
    private static int ALPHABET_SIZE = 26;
    Node<T> root;

    public Tri( ) {
        root = new Node( );
    }

    public T get( String key ) {

        if ( key == null ) {
            return null;
        }

        key = clean( key );

        Node current = root;
        for( int i = 0; i < key.length( ); i++ ) {
            int index = hash( key.charAt( i ) );
            Node node = current.getChild( index );

            if ( node == null ) {
                return null;
            }
            current = node;
        }
        return (T) current.getValue( );
    }

    public void insert( String key, T value ) {
        key = clean( key );

        Node current = root;
        for( int i = 0; i < key.length( ); i++ ) {
            int index = hash( key.charAt( i ) );
            Node node = current.getChild( index );

            if ( node == null ) {
                node = new Node( );
                current.setChild( index, node );
            }
            current = node;
        }
        current.setValue( value );
    }

    private String clean( String key ) {
        return key.toLowerCase().replaceAll("[^a-z]", "");
    }

    private int hash( char c ) {
        return c - OFFSET;
    }

    private char toChar( int index ) {
        return (char) ( index + OFFSET );
    }

    private class Node<T> {
        Node[] child = new Node[ ALPHABET_SIZE ];
        T value;

        public void setValue( T value ) {
            this.value = value;
        }

        public T getValue( ) {
            return value;
        }

        public Node getChild( int index ) {
            return child[ index ];
        }

        public void setChild( int index, Node node ) {
            child[ index ] = node;
        }
    }
}

File: src/main/kotlin/javabot/operations/time/Tri.kt
package javabot.operations.time

class Tri<T> {
    private var root: Node<T> = Node()

    fun get(key: String): T? {

        var cleaned = clean(key)

        var current = root
        for (i in 0..cleaned.length - 1) {
            val index = hash(cleaned[i])
            val node = current.getChild(index) ?: return null

            current = node
        }
        @Suppress("UNCHECKED_CAST")
        return current.value as T
    }

    fun insert(key: String, value: T) {
        var cleaned = clean(key)

        var current = root
        for (i in 0..cleaned.length - 1) {
            val index = hash(cleaned[i])
            var node = current.getChild(index)

            if (node == null) {
                node = Node()
                current.setChild(index, node)
            }
            current = node
        }
        current.value = value
    }

    private fun clean(key: String): String {
        return key.lowercase().replace("[^a-z]".toRegex(), "")
    }

    private fun hash(c: Char): Int {
        return c.code - OFFSET
    }

    private inner class Node<T> {
        var child = arrayOfNulls<Node<T>>(26)
        var value: T? = null

        fun getChild(index: Int): Node<T>? {
            return child[index]
        }

        fun setChild(index: Int, node: Node<T>) {
            child[index] = node
        }
    }

    companion object {
        private val OFFSET = 'a'.code
    }
}
